doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+  In general, it is not easy to predict whether GHC will inline something or not.
+
+The lack of consistency makes it difficult to reason about the efficiency of Plutus Tx programs, and difficult to write useful libraries for Plutus Tx that guarantee to have the desirable behaviors (e.g., short-circuiting).
+
+This ADR proposes a few changes to the Plutus Tx language and the compiler to make the semantics more predictable.
+It also proposes changes to the Plutus Tx standard library to make it more performant.
+
+## Proposed Changes
+
+Plutus Tx should be a strict-by-default language.
+Non-strict-by-default languages are rarely used in practice since non-strict evaluation leads to repeated evaluations.
+Unless we implement lazy evaluation for PLC in the future, strictness should be the default for Plutus Tx.
+Given this, the following changes are proposed.
+
+### Making Let-bindings Strict By Default
+
+In a strict-by-default language, it would make sense that let-bindings be strict by default, since that's what almost all strict languages do; plus, non-strict bindings should be used cautiously since they can easily lead to unbounded amount of repeated evaluation.
+But if make let-bindings strict by default, then the question is, how does one make a non-strict let-binding?
+
+Haskell supports lazy patterns with the `~` operator, and it can be used on let-bindings, e.g., `let ~x = rhs`. So one possibility is to make non-strict let-bindings this way.
+However, lazy pattern has no effect on simple pattern bindings (i.e., the pattern is a single variable), and it makes no difference in GHC Core, which means the Plutus Tx compiler can't actually see the `~`.
+It is technically possible to make `~` work but it would require adding a GHC parser plugin to capture `~` in GHC Core, among other things, adding a great deal of complexity.
+
+This is the reason why let-bindings have been non-strict by default in Plutus Tx, and creating strict let-bindings requires using bang patterns.
+
+However, we can do it in a different way.
+Although we can't modify Haskell's syntax, we can introduce functions to be handled specially by the Plutus Tx compiler.
+Specifically, we can introduce the following function:
+
+```haskell
+nonstrict :: forall a. a -> a
+nonstrict = id
+{-# NOINLINE nonstrict #-}
+```
+
+Then, to create a non-strict let-binding, we write
+
+```haskell
+let x = nonstrict rhs in body
+```
+
+Because of the NOINLINE pragma, `nonstrict` is guaranteed to be visible to the Plutus Tx compiler.
+Whenever the Plutus Tx compiler sees `let x = nonstrict rhs`, it first compiles `rhs` normally; then it creates a nonstrict binding in PIR: `let ~x = rhs_compiled`.
+
+What if `x` gets inlined by GHC, i.e., GHC turns `let x = nonstrict rhs in ...x...` into `...(nonstrict rhs)...`?
+That's not a problem, because if we see a `nonstrict rhs` that isn't the RHS of a binding, we create a new, non-strict binding from it, i.e., it becomes, in PIR,
```

Comment:
It's a shame we can't just use this strategy unconditionally, i.e. it would be nice if we could turn
```
let x = nonstrict rhs in body
```
into
```
let ~fresh = rhs
in let !x = fresh
in body
```
but that doesn't actually work. Anyway, we can discuss the specific details elsewhere. In fact, I think probably this ADR shouldn't cover them: the _decision_ is to handle this through explicit signalling that indicates a non-strict binding, and how we implement that isn't so relevant.

Reply:
In fact, I think this behaviour of `nonstrict` is _actually pointless_. There is no difference between `... t ...` and `let ~x = t in ... x ...`. In both cases we are going to evaluate `t` exactly when we hit the location where `t` is originally. So standalone `nonstrict` is pointless.

TBH I still don't love this approach. I don't like that we're putting something on the RHS of the binding to affect the LHS, and that it doesn't really have a standalone meaning or work anywhere else. I wish we could use an annotation or something, but source annotations can't be used on nested bindings :(